cocos/animation/marionette/blend-2d.ts

```diff
@@ -256,50 +266,189 @@ const getGradientBandCartesianCoords: GetGradientBandCoords = (pI, pJ, input, pI
     Vec2.subtract(pIpJ, pJ, pI);
 };
 
-const getGradientBandPolarCoords = ((): GetGradientBandCoords => {
-    const axis = new Vec3(0, 0, 0); // buffer for axis
-    const tmpV3 = new Vec3(0, 0, 0); // buffer for temp vec3
-    const pQueriedProjected = new Vec3(0, 0, 0); // buffer for pQueriedProjected
-    const pi3 = new Vec3(0, 0, 0); // buffer for pi3
-    const pj3 = new Vec3(0, 0, 0); // buffer for pj3
-    const pQueried3 = new Vec3(0, 0, 0); // buffer for pQueried3
-    return (pI, pJ, input, pIpInput, pIpJ) => {
-        let aIJ = 0.0;
-        let aIQ = 0.0;
-        let angleMultiplier = 2.0;
-        Vec3.set(pQueriedProjected, input.x, input.y, 0.0);
-        if (Vec2.equals(pI, Vec2.ZERO)) {
-            aIJ = Vec2.angle(input, pJ);
-            aIQ = 0.0;
-            angleMultiplier = 1.0;
-        } else if (Vec2.equals(pJ, Vec2.ZERO)) {
-            aIJ = Vec2.angle(input, pI);
-            aIQ = aIJ;
-            angleMultiplier = 1.0;
+const PRECOMPUTED_VIJ_DATA_STRIDE = 3;
+
+export class PolarSpaceGradientBandInterpolator2D {
+    private static _CACHE_INPUT_DIRECTION = new Vec2();
+
+    private static _CACHE_VIJ = new Vec2();
+
+    private static _CACHE_VIX = new Vec2();
+
+    private static _ANGLE_MULTIPLIER = 1.0;
+
+    constructor (examples: readonly Readonly<Vec2>[]) {
+        const {
+            _ANGLE_MULTIPLIER: angleMultiplier,
+        } = PolarSpaceGradientBandInterpolator2D;
+
+        const nExamples = examples.length;
+
+        const exampleMagnitudes = this._exampleMagnitudes = new Array<number>(nExamples).fill(0.0);
+        const exampleDirections = this._exampleDirections = examples.map((example, iExample) => {
+            const direction = Vec2.copy(new Vec2(), example);
+            const magnitude = Vec2.len(direction);
+            exampleMagnitudes[iExample] = magnitude;
+            if (!approx(magnitude, 0.0, 1e-5)) {
+                Vec2.multiplyScalar(direction, direction, 1.0 / magnitude);
+            }
+            return direction;
+        });
+
+        const precomputedPIJs = this._precomputedPIJs = new Float32Array(PRECOMPUTED_VIJ_DATA_STRIDE * nExamples * nExamples);
```

Comment:
What does `PIJ` means?

Reply:
I've documented it in https://github.com/cocos/cocos-engine/pull/14203/commits/325ebdfe0bc48e6f04be66f8ee062066296437c7 .